### PR TITLE
TestFramework: Add Python Miniscript Support

### PR DESCRIPTION
**Pull request by jachiang**
This PR introduces basic Miniscript support to the TestFramework. 

I believe this would be useful for:

- Functional tests with more complex script satisfaction(s)
- Functional tests for future wallet RPC calls which handle Miniscript
- Entry point for developers looking to integrate Miniscript into external wallet projects (ecdsa, SomberNight, dgpv).

**Usage:**

Miniscript string constructor.
```
# key, delay in descriptor below must be represented as hex/int.

example_desc = “or_b(or_i(n:thresh_m(k,key,key),0),a:or_i(0,older(delay)))”

miniscript_node = Node.from_desc(example_desc)
```

Miniscript CScript constructor.
```
miniscript_node = Node.from_script(cscript)
```

Miniscript type and properties.
```
miniscript_node.p.to_string()
```

(Dis)satisfaction methods, returns list of tuple-lists.
```
# Canonical (dis)satisfying witnesses
miniscript_node.sat
miniscript_node.dsat

# Non-canonical (dis)satisfying witnesses
miniscript_node.sat_ncan
miniscript_node.dsat_ncan
```

Each tuple-list encodes a single, unique satisfying witness:
```
[
	(SatType, Value),
	(SatType, Value),
	(SatType, Value),
	…
]
```
The SatType, Value tuples encode the following witness information in correct order. Exception: Time/Delay tuples are not witness elements and will always be positioned at the beginning of the list.

|  SatType | Value | 
|---|---|
| OLDER | Delay int | 
| AFTER | Time int |
| SIGNATURE | 33B Key/20B HASH160 Digest  |
| KEY_AND_HASH160_PREIMAGE | 20B HASH160 Digest |
| SHA256_PREIMAGE | 32B SHA256 Digest |
| HASH256_PREIMAGE | 32B HASH256 Digest |
| RIPEMD160_PREIMAGE |  20B RIPEMD160 Digest |
| HASH160_PREIMAGE | 20B HASH160 Digest |
| DATA | Bytes |

Our example `miniscript_node.sat` above returns the following 2 satisfying witnesses:
```
[
	[
		(<SatType.DATA: 8>, b'\x01'), 
		(<SatType.DATA: 8>, b''), 
		(<SatType.SIGNATURE: 2>,  key_bytes), 
		(<SatType.SIGNATURE: 2>,  key_bytes), 
		(<SatType.DATA: 8>, b'\x01')
	], 
	[
		(<SatType.OLDER: 0>, 45), 
		(<SatType.DATA: 8>, b''), 
		(<SatType.DATA: 8>, b'')
	]
]
```

**Tests:**

I have included tests in feature_miniscript.py to facilitate potential reviews, but would suggest to omit these if the maintainers decide to merge this PR.

### CODE REVIEW

**Comment by instagibbs** [comment]
@sanket1729 @apoelstra may be interested too

File: test/functional/test_framework/miniscript.py
Comment (dgpv) [inline test/functional/test_framework/miniscript.py]: this is a duplicate of the code block above

File: test/functional/test_framework/miniscript.py
Comment (dgpv) [inline test/functional/test_framework/miniscript.py]: should be `=` rather than `+=` in two lines above, otherwise `child_is_e` will be converted to in tan will always be `>= 1`

File: test/functional/test_framework/miniscript.py
Comment (dgpv) [inline test/functional/test_framework/miniscript.py]: why not `child.t.name.startswith('WRAP_')` ?

Comment (jachiang) [reply test/functional/test_framework/miniscript.py]: That's much better. Thx.

File: test/functional/test_framework/miniscript.py
Comment (jachiang) [reply test/functional/test_framework/miniscript.py]: Excellent catch. Thx again.

**Comment by michaelfolkson** [comment]
This looks great. To get the concept ACK I'm wondering what the arguments against are. I think (I may be mistaken) these are broadly long term motivations and sketched out roadmap-like issues.

1) Does Bitcoin Core *need* to be able to recognize Miniscript in the future? Are the future upsides sufficiently material to introduce it? Presumably Miniscript could always be translated to Script externally to Core at little cost?

2) Is this a stepping stone towards greater support for Miniscript (and possibly a policy language too) in Core? What impact does this have on the use and ongoing development of external libraries like the [Rust Miniscript library](https://github.com/apoelstra/rust-miniscript)?

3) I'm not convinced of the rationale "entry point for developers" as a motivation. Does Miniscript support in the test framework provide any material assistance/assurance to external wallet developers? 

I'm excited by the potential of Miniscript and look forward to following future work on this.

File: test/functional/feature_miniscript.py
Comment (emilengler) [inline test/functional/feature_miniscript.py]: Please change this to "The Bitcoin Core developers" or add your name on line 105 in https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/copyright_header.py

Otherwise it will get overridden by the copyright-headers script

Comment (jachiang) [reply test/functional/feature_miniscript.py]: Thank you for the review and tip, will gladly amend copyright header as you suggested!

**Comment by practicalswift** [comment]
@jachiang

Very nice to see another Miniscript implementation! I love Miniscript! :)

You might want to check your Python implementation for the following issues I've reported in the C++ and Rust implementations:
* [Issues reported in the C++ implementation (7 of which I would consider bugs)](https://github.com/sipa/miniscript/issues?utf8=%E2%9C%93&q=is%3Aissue+author%3Apracticalswift+)
* [Issues reported in the Rust implementation (8 of which I would consider bugs)](https://github.com/apoelstra/rust-miniscript/issues?utf8=%E2%9C%93&q=is%3Aissue+author%3Apracticalswift+)

The bugs are somewhat overlapping across implementations.

As you can see six of the seven bugs in the Rust implementation have been resolved, and hopefully the eight bugs in the C++ implementation will be resolved soon :)

Of the eight bugs in the C++ implementation only one bug affects the initial Bitcoin Core Miniscript PR #16800 ("Basic Miniscript support in output descriptors"). That is a heap out-of-bounds read in `Node::CalcOps` (called indirectly from `miniscript::FromScript`) which I reported in https://github.com/bitcoin/bitcoin/pull/16800#issuecomment-530808234 and suggested a solution for in https://github.com/sipa/miniscript/pull/18 ("Avoid heap out-of-bounds read in Node::CalcOps (test case: OP_0 OP_2 OP_EQUAL) and assertion failure in ComputeType (test case: OP_0 OP_0 OP_EQUAL)").

**Comment by michaelfolkson** [comment]
Concept ACK. Moving these high level questions to a RFC (#18040) so as not to hold up review of the details of this PR. https://github.com/bitcoin/bitcoin/pull/17975#issuecomment-577095603

File: test/functional/test_framework/miniscript.py
Comment (sanket1729) [inline test/functional/test_framework/miniscript.py]: I think this should be ` prop_str += "f" if child_x.p.s or child_y.p.f else "" ` . Since the left child must have a base type`V`, it always has malleability property `f`. The above statement will make all `and_v` has property `f`. 
There is only one possible non-canonical dissat being [sat(X), dissat(Y)] which gives us self.f = `sx or fy`. 

One valid miniscript counter-example: `and_v(vc:pk(A),and_v(v:after(10),sha256(H)))` where right branch of first `and_v` is not an `f`.

**Review by sanket1729 — CHANGES_REQUESTED** [review]
Great work with the PR :) . I just a partial review, will do a more detailed review over the next couple of days.

File: test/functional/test_framework/miniscript.py
Comment (sanket1729) [inline test/functional/test_framework/miniscript.py]: Should be if "Bdu" if is_B, otherwise ""

**Review by sanket1729 — CHANGES_REQUESTED** [review]
Changes requested according to revised typing rules in http://bitcoin.sipa.be/miniscript/
Reviewed typing rules. I will review the sat/dsat and script parsing in the following week.

File: test/functional/test_framework/miniscript.py
Comment (jachiang) [reply test/functional/test_framework/miniscript.py]: Yes absolutely! Thanks for the correction!

File: test/functional/test_framework/miniscript.py
Comment (jachiang) [reply test/functional/test_framework/miniscript.py]: Ah yes, both V are from revised rules.

File: test/functional/test_framework/miniscript.py
Comment (jachiang) [reply test/functional/test_framework/miniscript.py]: Got it, revised rules again :)

File: test/functional/test_framework/miniscript.py
Comment (jachiang) [reply test/functional/test_framework/miniscript.py]: This was a mistake in encoding the property rules, thank you.

**Comment by laanwj** [comment]
Concept ACK

**Comment by jachiang** [comment]
Apologies for the delay updating this PR. 

f8a30b4c7 includes rebase and changes suggested by reviewers so far.

File: test/functional/test_framework/miniscript.py
Comment (JeremyRubin) [inline test/functional/test_framework/miniscript.py:1065]: Should be type "V" if both "V".

**Review by darosior — COMMENTED** [review]
Concept ACK. Didn't review tests nor satisfaction yet. Will do post rebase.

As mentioned elsewhere, i'm working on a [python-miniscript](https://github.com/darosior/python-miniscript) package started from your branch. All my comments about updates and corrections can be found [as commits there](https://github.com/darosior/python-miniscript/commits/master). I also included more test cases and changed the structure + nits for my Python OCD, of course feel free to take anything from there.

Hopefully Github will keep my comments ordered by commits, but the main points are:
- You are missing a lot of child property checks when constructing nodes. 
- I think you should get rid of the `x` property
- Miniscript has been updated, `pk` is now `pk_k` and `pk` is an alias to `c:pk_k`, `pkh` is now an alias to `c:pk_h`, and `thresh_m` was renamed to `multi`. Some new properties were also added for timelocks conflicts, but they are not yet merged in both implementations so i didn't comment on it.

**Comment by JeremyRubin** [comment]
Another option here which might not be too bad in the scheme of things would be to check in a WASM blob that is gitian built of e.g. rust miniscript, and then call it using a wasm runtime https://github.com/wasmerio/wasmer-python

**Comment by jachiang** [comment]
> the main benefit being we don't have to check in yet another implementation of miniscript to maintain as miniscript evolves, we can use the rust or c++ one today, and when the C++ becomes merged & available as an RPC we can drop it.

In the absence of a formal, (mechanizable) specification of Miniscript (MS), separate implementations have higher chances of finding bugs or problems than a single, canonical one. I believe this should be a goal of the test-framework. Having said that, I believe the biggest challenge here is the lack of a formal specification of MS and by extension, Script itself, without which, the correctness of the MS type-system is impossible to verify. One must assume the MS authors are working on this.

**Comment by Sjors** [comment]
> In the absence of a formal, (mechanizable) specification of Miniscript (MS), separate implementations have higher chances of finding bugs or problems than a single, canonical one. I think this should be a goal of the test-framework.

Agreed. Concept ACK.

**Comment by michaelfolkson** [comment]
I'm assuming this PR could be closed then @darosior? Direct interested parties for Python Miniscript/Miniscript support in test framework to https://github.com/darosior/python-bip380? When that's in a ready state a new PR will be opened in this repo?

**Comment by michaelfolkson** [comment]
> @michaelfolkson it's not my PR to close

Ok up to @jachiang then. Your repo is building on his work and he should be credited when a new PR is opened (which I'm sure you would anyway)

**Comment by Sjors** [comment]
> @michaelfolkson it's not my PR to close. 😅
> 
> > When that's in a ready state a new PR will be opened in this repo?
> 
> I think having multiple implementations of Miniscript is good, but having multiple ones _as part of this repo_ would only have marginal benefits, if at all.

A good set of test vectors is one way to compare implementations. Lacking that, it's nice if we can generate a few thousand miniscripts in Python and compare them with what the C++ code does.

Now that #24148 has landed, it should be more straight forward to rebase this PR and do the comparison by e.g. calling `deriveaddresses` on various `wsh(some_miniscript)`.

**Comment by darosior** [comment]
We did have those, but for some reason we dropped them in the C++ test framework. We can probably add those back. Dmitri generated thousands of Miniscripts (for testing timelock mixes initially) we tested rust-miniscript and the C++ implem against.

If you want an up to dare Python implem i have one at https://github.com/darosior/python-bip380. But again i'm not convinced we should maintain it as part of this repo.
-------- Original Message --------

**Comment by achow101** [comment]
Closing this as it has not had any activity in a while. If you are interested in continuing work on this, please leave a comment so that it can be reopened.

**Pull request closed without merging**
